### ISSUE TITLE
Updating cluster agentEnvVars does not trigger system agent upgrade

### ISSUE THREAD

**Issue by gitlawr** [rancher/rancher#34908]
**Rancher Server Setup**
- Rancher version: v2.6.0
- Installation option: Embedded in Harvester v0.3.0-preview1, with rancherd + rke2

**Describe the bug**

**To Reproduce**
1. Install Harvester v0.3.0-preview1
2. Edit rancher-webhook image tag to v0.2.1-rc3 to workaround the issue https://github.com/rancher/rancher/issues/34691
3. Run the following command and add/update the spec.agentEnvVars field:
```
kubectl edit clusters.provisioning.cattle.io  local -n fleet-local
### example below
spec:
  agentEnvVars:
  - name: NO_PROXY
    value: localhost
```

**Result**
1. No system agent pods get created to propagate the envvars.
2. The envvars are not applied to RKE2 services.

**Expected Result**
1. System agent pods are deployed for rke2 upgrade.
2. The configured envvars show up in `/etc/systemd/system/rancher-system-agent.env` on each rke2 node.
3. RKE2 services get started to apply updated envvars.

**Additional context**
1. The envvar is set in the SUC plan, can be checked by `kubectl get plan system-agent-upgrader -o yaml -n cattle-system`
2. If I remove the `plan.upgrade.cattle.io/system-agent-upgrader` label on a node, a system agent pod will be created and the envvars are propagated to the `rancher-system-agent.env` file on the node.

**Comment by StrongMonkey** [comment]
According to @ibuildthecloud This is a release blocker of harvester.

**Comment by thedadams** [comment]
See this comment for the testing strategy: https://github.com/rancher/rancher/issues/32633#issuecomment-928287779

**Comment by gitlawr** [comment]
With v2.6.1-rc5:

Install Harvester, run `kubectl edit clusters.provisioning.cattle.io  local -n fleet-local`

Edit and save:
```
  agentEnvVars:
  - name: HTTP_PROXY
    value: http://192.168.0.1:3888
  - name: HTTPS_PROXY
    value: http://192.168.0.1:3888
  - name: NO_PROXY
    value: localhost,127.0.0.1,0.0.0.0,10.0.0.0/8,cattle-system.svc,192.168.0.0/16,172.16.0.0/16,.svc,.cluster.local
```

Result:
A apply-system-agent-upgrader pod runs but never completes. Logs:
```
+ TMPDIRBASE=/var/lib/rancher/agent/tmp
+ mkdir -p /host/var/lib/rancher/agent/tmp
+ chroot /host /bin/sh -c 'mktemp -d -p /var/lib/rancher/agent/tmp'
+ TMPDIR=/var/lib/rancher/agent/tmp/tmp.R2JIT7M9ku
+ cp /opt/rancher-system-agent-suc/install.sh /host/var/lib/rancher/agent/tmp/tmp.R2JIT7M9ku
+ cp /opt/rancher-system-agent-suc/rancher-system-agent /host/var/lib/rancher/agent/tmp/tmp.R2JIT7M9ku
+ chmod +x /host/var/lib/rancher/agent/tmp/tmp.R2JIT7M9ku/install.sh
+ '[' -n tt ]
+ NODE_FILE=/host/var/lib/rancher/agent/tmp/tmp.R2JIT7M9ku/node.yaml
+ kubectl get node tt -o yaml
+ '[' -z  ]
+ grep -q 'node-role.kubernetes.io/etcd: "true"' /host/var/lib/rancher/agent/tmp/tmp.R2JIT7M9ku/node.yaml
+ export 'CATTLE_ROLE_ETCD=true'
+ '[' -z  ]
+ grep -q 'node-role.kubernetes.io/controlplane: "true"' /host/var/lib/rancher/agent/tmp/tmp.R2JIT7M9ku/node.yaml
+ '[' -z  ]
+ grep -q 'node-role.kubernetes.io/control-plane: "true"' /host/var/lib/rancher/agent/tmp/tmp.R2JIT7M9ku/node.yaml
+ export 'CATTLE_ROLE_CONTROLPLANE=true'
+ '[' -z true ]
+ export 'CATTLE_AGENT_BINARY_LOCAL=true'
+ export 'CATTLE_AGENT_BINARY_LOCAL_LOCATION=/var/lib/rancher/agent/tmp/tmp.R2JIT7M9ku/rancher-system-agent'
+ '[' -s /host/etc/systemd/system/rancher-system-agent.env ]
+ chroot /host /var/lib/rancher/agent/tmp/tmp.R2JIT7M9ku/install.sh
[INFO]  Using default agent configuration directory /etc/rancher/agent
[INFO]  Using default agent var directory /var/lib/rancher/agent
[WARN]  /usr/local is read-only or a mount point; installing to /opt/rancher-system-agent
[ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
[ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
[ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
[ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
[ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
```

To debug, exec to the pod and run the same curl command 
```
chroot /host
curl --connect-timeout 60 --max-time 60 --write-out '%{ssl_verify_result}\n' -v -fL https://10.53.221.44/healthz

* Uses proxy env variable NO_PROXY == 'localhost,127.0.0.1,0.0.0.0,10.0.0.0/8,cattle-system.svc,192.168.0.0/16,172.16.0.0/16,.svc,.cluster.local'
* Uses proxy env variable HTTPS_PROXY == 'http://192.168.0.1:3888'
*   Trying 192.168.0.1:3888...
* TCP_NODELAY set
* Connected to 192.168.0.1 (192.168.0.1) port 3888 (#0)
* allocate connect buffer!
* Establish HTTP proxy tunnel to 10.53.221.44:443
> CONNECT 10.53.221.44:443 HTTP/1.1
> Host: 10.53.221.44:443
> User-Agent: curl/7.66.0
> Proxy-Connection: Keep-Alive
>
^C
```

The cause is that curl command does not support using CIDR notation in NO_PROXY as suggested in https://rancher.com/docs/rancher/v2.6/en/installation/other-installation-methods/single-node-docker/proxy/

We may consider using `curl --noproxy '*'` given that it's the internal Rancher service.

**Comment by thedadams** [comment]
The suggested solution won't work unless Rancher can be contacted without the proxy. It seems that it can in the case you have provided, but this won't be the case all the time. I will continue to look for an acceptable solution.

**Comment by yasker** [comment]
Due to being unable to fix this issue for Rancher v2.6.1, Harvester v0.3.0 won't be able to support Air Gap environment. Air Gap Environment will be supported in Harvester v1.0.0.

**Comment by gitlawr** [comment]
Verified in v2.6-931a0f09d0f85ade0dcaa8fbe9d6e0c58ae52bbc-head
After adding the proxy environment variables, the apply-system-agent-upgrader pods successfully completed.
The environment variables are present in `/var/lib/rancher/rke2/system-agent-installer/rke2-sa.env` on each node.
I can run a deployment with `nginx:latest` image as expected.
